File: example/Lab05_3.py
from Crypto.PublicKey import RSA
from Crypto.Signature import pkcs1_15
from Crypto.Hash import SHA1
from Crypto.Cipher import DES, PKCS1_OAEP
from Crypto.Util.Padding import pad, unpad
from Crypto.Random import get_random_bytes
import os
import json
import base64


class DigitalSignature:
    def __init__(self):
        self.private_key = None
        self.public_key = None
        self.des_key = None
    
    def generate_rsa_keys(self, key_size=2048):
        """Tạo cặp khóa RSA"""
        print(f"Tạo cặp khóa RSA {key_size} bits...")
        key_pair = RSA.generate(key_size)
        self.private_key = key_pair
        self.public_key = key_pair.publickey()
        print("Tạo khóa thành công!")
    
    def save_rsa_keys(self, base_filename):
        """Lưu khóa RSA vào file"""
        with open(f"{base_filename}.key", 'wb') as f:
            f.write(self.private_key.export_key('PEM'))
        with open(f"{base_filename}.pub", 'wb') as f:
            f.write(self.public_key.export_key('PEM'))
        print(f"Đã lưu khóa RSA:")
        print(f"  Private: {base_filename}.key")
        print(f"  Public: {base_filename}.pub")
    
    def load_private_key(self, filename):
        """Tải private key từ file"""
        with open(filename, 'rb') as f:
            self.private_key = RSA.import_key(f.read())
        print(f"Đã tải private key từ {filename}")
    
    def load_public_key(self, filename):
        """Tải public key từ file"""
        with open(filename, 'rb') as f:
            self.public_key = RSA.import_key(f.read())
        print(f"Đã tải public key từ {filename}")
    
    def set_des_key(self, key):
        """
        Thiết lập khóa DES (8 bytes)
        Giả sử khóa này đã được chia sẻ giữa bên gửi và bên nhận
        """
        if isinstance(key, str):
            key = key.encode()
        if len(key) != 8:
            raise ValueError("DES key phải có độ dài 8 bytes")
        self.des_key = key
    
    def sign(self, data):
        """
        Ký trên một thông điệp bằng SHA1withRSA
        
        Args:
            data: Dữ liệu cần ký (string hoặc bytes)
        
        Returns:
            signature: Chữ ký số (bytes)
        """
        if self.private_key is None:
            raise ValueError("Chưa có private key!")
        
        if isinstance(data, str):
            data = data.encode('utf-8')
        
        hash_obj = SHA1.new(data)
        
        signature = pkcs1_15.new(self.private_key).sign(hash_obj)
        
        print(f"Đã ký thông điệp (độ dài: {len(data)} bytes)")
        print(f"Chữ ký (độ dài: {len(signature)} bytes)")
        
        return signature
    
    def verify_signature(self, data, signature):
        if self.public_key is None:
            raise ValueError("Chưa có public key!")
        
        if isinstance(data, str):
            data = data.encode('utf-8')
        
        hash_obj = SHA1.new(data)
        
        try:
            pkcs1_15.new(self.public_key).verify(hash_obj, signature)
            print("✓ Chữ ký hợp lệ!")
            return True
        except (ValueError, TypeError):
            print("✗ Chữ ký không hợp lệ!")
            return False
    
    def sign_and_encrypt(self, data, output_file):
        if self.private_key is None:
            raise ValueError("Chưa có private key!")
        if self.des_key is None:
            raise ValueError("Chưa thiết lập DES key!")
        
        print("\n=== SIGN AND ENCRYPT ===")
        
        # Chuyển data thành bytes
        if isinstance(data, str):
            data = data.encode('utf-8')
        
        # Bước 1: Ký dữ liệu
        print("Bước 1: Ký dữ liệu...")
        signature = self.sign(data)
        
        # Bước 2: Tạo gói dữ liệu (data + signature)
        print("Bước 2: Tạo gói dữ liệu...")
        package = {
            'data_length': len(data),
            'signature_length': len(signature),
            'data': base64.b64encode(data).decode('utf-8'),
            'signature': base64.b64encode(signature).decode('utf-8')
        }
        package_json = json.dumps(package)
        package_bytes = package_json.encode('utf-8')
        
        # Bước 3: Mã hóa toàn bộ bằng DES
        print("Bước 3: Mã hóa toàn bộ bằng DES...")
        cipher = DES.new(self.des_key, DES.MODE_CBC)
        iv = cipher.iv
        
        # Padding và mã hóa
        padded_data = pad(package_bytes, DES.block_size)
        ciphertext = cipher.encrypt(padded_data)
        
        # Lưu vào file (IV + ciphertext)
        with open(output_file, 'wb') as f:
            f.write(iv)
            f.write(ciphertext)
        
        print(f"✓ Đã ký và mã hóa, lưu vào {output_file}")
        print(f"  - Dữ liệu gốc: {len(data)} bytes")
        print(f"  - Chữ ký: {len(signature)} bytes")
        print(f"  - File mã hóa: {len(iv) + len(ciphertext)} bytes")
        
        return {
            'data_length': len(data),
            'signature_length': len(signature),
            'encrypted_length': len(iv) + len(ciphertext)
        }
    
    def decrypt_and_verify(self, input_file):
        if self.public_key is None:
            raise ValueError("Chưa có public key!")
        if self.des_key is None:
            raise ValueError("Chưa thiết lập DES key!")
        
        print("\n=== DECRYPT AND VERIFY ===")
        
        # Bước 1: Đọc file và giải mã bằng DES
        print("Bước 1: Giải mã bằng DES...")
        with open(input_file, 'rb') as f:
            iv = f.read(8)
            ciphertext = f.read()
        
        cipher = DES.new(self.des_key, DES.MODE_CBC, iv=iv)
        padded_data = cipher.decrypt(ciphertext)
        package_bytes = unpad(padded_data, DES.block_size)
        
        # Bước 2: Parse JSON để lấy data và signature
        print("Bước 2: Trích xuất dữ liệu và chữ ký...")
        package_json = package_bytes.decode('utf-8')
        package = json.loads(package_json)
        
        data = base64.b64decode(package['data'])
        signature = base64.b64decode(package['signature'])
        
        print(f"  - Dữ liệu: {len(data)} bytes")
        print(f"  - Chữ ký: {len(signature)} bytes")
        
        # Bước 3: Xác minh chữ ký
        print("Bước 3: Xác minh chữ ký...")
        is_valid = self.verify_signature(data, signature)
        
        return data, is_valid


def demo_basic_sign_verify():
    """Demo cơ bản về ký và xác minh"""
    print("\n" + "=" * 60)
    print("DEMO: KÝ VÀ XÁC MINH CHỮ KÝ SỐ")
    print("=" * 60)
    
    ds = DigitalSignature()
    
    # Tạo khóa
    ds.generate_rsa_keys(2048)
    
    # Thông điệp cần ký
    message = "Đây là thông điệp cần được ký và xác minh!"
    print(f"\nThông điệp: {message}")
    
    # Ký thông điệp
    print("\n--- Ký thông điệp ---")
    signature = ds.sign(message)
    
    # Xác minh chữ ký (đúng)
    print("\n--- Xác minh chữ ký (thông điệp gốc) ---")
    ds.verify_signature(message, signature)
    
    # Xác minh chữ ký (sai - thông điệp bị sửa)
    print("\n--- Xác minh chữ ký (thông điệp bị sửa) ---")
    tampered_message = "Đây là thông điệp ĐÃ BỊ SỬA ĐỔI!"
    ds.verify_signature(tampered_message, signature)


def demo_sign_and_encrypt():
    """Demo ký và mã hóa toàn bộ"""
    print("\n" + "=" * 60)
    print("DEMO: KÝ VÀ MÃ HÓA TOÀN BỘ")
    print("=" * 60)
    
    # Bên gửi
    print("\n--- BÊN GỬI ---")
    sender = DigitalSignature()
    sender.generate_rsa_keys(2048)
    sender.set_des_key("sharedKY")  # Khóa DES đã chia sẻ
    
    message = "Thông điệp bí mật cần được bảo mật và xác thực!"
    print(f"Thông điệp: {message}")
    
    # Ký và mã hóa
    sender.sign_and_encrypt(message, "encrypted_message.bin")
    
    # Bên nhận
    print("\n--- BÊN NHẬN ---")
    receiver = DigitalSignature()
    receiver.public_key = sender.public_key  # Nhận public key từ bên gửi
    receiver.set_des_key("sharedKY")  # Khóa DES đã chia sẻ
    
    # Giải mã và xác minh
    data, is_valid = receiver.decrypt_and_verify("encrypted_message.bin")
    
    print(f"\n--- KẾT QUẢ ---")
    print(f"Dữ liệu nhận được: {data.decode('utf-8')}")
    print(f"Chữ ký hợp lệ: {is_valid}")
    
    # Dọn dẹp
    os.remove("encrypted_message.bin")


def main():
    print("=" * 60)
    print("CHƯƠNG TRÌNH KÝ VÀ XÁC MINH CHỮ KÝ SỐ")
    print("=" * 60)
    
    print("\nChọn chức năng:")
    print("1. Tạo cặp khóa RSA")
    print("2. Ký thông điệp")
    print("3. Xác minh chữ ký")
    print("4. Ký và mã hóa toàn bộ (signAndEncrypt)")
    print("5. Giải mã và xác minh (decryptAndVerify)")
    print("6. Demo cơ bản")
    print("7. Demo ký và mã hóa")
    
    choice = input("\nLựa chọn (1-7): ").strip()
    
    ds = DigitalSignature()
    
    if choice == '1':
        # Tạo khóa
        key_size = input("Nhập kích thước khóa (mặc định 2048): ").strip()
        key_size = int(key_size) if key_size else 2048
        
        ds.generate_rsa_keys(key_size)
        
        base_name = input("Nhập tên cơ sở cho file khóa: ").strip()
        if not base_name:
            base_name = "signature_key"
        
        ds.save_rsa_keys(base_name)
    
    elif choice == '2':
        # Ký thông điệp
        priv_key = input("Nhập tên file private key: ").strip()
        if not os.path.exists(priv_key):
            print(f"File {priv_key} không tồn tại!")
            return
        
        ds.load_private_key(priv_key)
        
        message = input("Nhập thông điệp cần ký: ").strip()
        signature = ds.sign(message)
        
        # Lưu chữ ký
        sig_file = input("Nhập tên file lưu chữ ký (mặc định: signature.bin): ").strip()
        if not sig_file:
            sig_file = "signature.bin"
        
        with open(sig_file, 'wb') as f:
            f.write(signature)
        print(f"Đã lưu chữ ký vào {sig_file}")
    
    elif choice == '3':
        # Xác minh chữ ký
        pub_key = input("Nhập tên file public key: ").strip()
        if not os.path.exists(pub_key):
            print(f"File {pub_key} không tồn tại!")
            return
        
        ds.load_public_key(pub_key)
        
        message = input("Nhập thông điệp gốc: ").strip()
        
        sig_file = input("Nhập tên file chứa chữ ký: ").strip()
        if not os.path.exists(sig_file):
            print(f"File {sig_file} không tồn tại!")
            return
        
        with open(sig_file, 'rb') as f:
            signature = f.read()
        
        ds.verify_signature(message, signature)
    
    elif choice == '4':
        # Ký và mã hóa
        priv_key = input("Nhập tên file private key: ").strip()
        if not os.path.exists(priv_key):
            print(f"File {priv_key} không tồn tại!")
            return
        
        ds.load_private_key(priv_key)
        
        des_key = input("Nhập khóa DES (8 ký tự): ").strip()
        if len(des_key) != 8:
            print("Khóa DES phải có độ dài 8 ký tự!")
            return
        
        ds.set_des_key(des_key)
        
        message = input("Nhập thông điệp: ").strip()
        output = input("Nhập tên file đầu ra (mặc định: encrypted.bin): ").strip()
        if not output:
            output = "encrypted.bin"
        
        ds.sign_and_encrypt(message, output)
    
    elif choice == '5':
        # Giải mã và xác minh
        pub_key = input("Nhập tên file public key: ").strip()
        if not os.path.exists(pub_key):
            print(f"File {pub_key} không tồn tại!")
            return
        
        ds.load_public_key(pub_key)
        
        des_key = input("Nhập khóa DES (8 ký tự): ").strip()
        if len(des_key) != 8:
            print("Khóa DES phải có độ dài 8 ký tự!")
            return
        
        ds.set_des_key(des_key)
        
        input_file = input("Nhập tên file cần giải mã: ").strip()
        if not os.path.exists(input_file):
            print(f"File {input_file} không tồn tại!")
            return
        
        data, is_valid = ds.decrypt_and_verify(input_file)
        
        print(f"\n--- KẾT QUẢ ---")
        print(f"Dữ liệu: {data.decode('utf-8')}")
        print(f"Chữ ký hợp lệ: {is_valid}")
    
    elif choice == '6':
        demo_basic_sign_verify()
    
    elif choice == '7':
        demo_sign_and_encrypt()
    
    else:
        print("Lựa chọn không hợp lệ!")


if __name__ == "__main__":
    main()

File: code/Lab05_3.java
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.*;
import java.security.spec.*;
import javax.crypto.*;
import javax.crypto.spec.*;
import java.util.Scanner;

/**
 * Lab05_3: Digital Signatures with SHA1withRSA
 * 
 * WARNING: SHA-1 is cryptographically broken (collision attacks demonstrated).
 * This implementation uses SHA1withRSA for EDUCATIONAL PURPOSES ONLY as required by assignment.
 * 
 * For production use:
 * - Use SHA256withRSA or SHA512withRSA instead
 * - Consider using EdDSA (Ed25519) for better performance and security
 * 
 * Implements:
 * 1. Basic sign/verify operations
 * 2. signAndEncrypt: Sign data, then encrypt (data + signature) with DES
 * 3. decryptAndVerify: Decrypt data, then verify signature
 */
public class Lab05_3 {
    
    private PrivateKey privateKey;
    private PublicKey publicKey;
    private byte[] desKey;
    
    /**
     * Generates RSA key pair for signatures
     */
    public void generateRSAKeys(int keySize) throws NoSuchAlgorithmException {
        System.out.println("Generating " + keySize + "-bit RSA key pair...");
        KeyPairGenerator keyGen = KeyPairGenerator.getInstance("RSA");
        keyGen.initialize(keySize);
        KeyPair keyPair = keyGen.generateKeyPair();
        
        privateKey = keyPair.getPrivate();
        publicKey = keyPair.getPublic();
        
        System.out.println("[SUCCESS] Key generation successful!");
    }
    
    /**
     * Saves RSA keys to files
     */
    public void saveRSAKeys(String baseFilename) throws IOException {
        try (FileOutputStream fos = new FileOutputStream(baseFilename + ".key")) {
            fos.write(privateKey.getEncoded());
        }
        
        try (FileOutputStream fos = new FileOutputStream(baseFilename + ".pub")) {
            fos.write(publicKey.getEncoded());
        }
        
        System.out.println("[SUCCESS] RSA keys saved:");
        System.out.println("  Private: " + baseFilename + ".key");
        System.out.println("  Public: " + baseFilename + ".pub");
    }
    
    /**
     * Loads private key from file
     */
    public void loadPrivateKey(String filename) throws Exception {
        byte[] keyBytes = Files.readAllBytes(Paths.get(filename));
        PKCS8EncodedKeySpec spec = new PKCS8EncodedKeySpec(keyBytes);
        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
        privateKey = keyFactory.generatePrivate(spec);
        System.out.println("Loaded private key from " + filename);
    }
    
    /**
     * Loads public key from file
     */
    public void loadPublicKey(String filename) throws Exception {
        byte[] keyBytes = Files.readAllBytes(Paths.get(filename));
        X509EncodedKeySpec spec = new X509EncodedKeySpec(keyBytes);
        KeyFactory keyFactory = KeyFactory.getInstance("RSA");
        publicKey = keyFactory.generatePublic(spec);
        System.out.println("Loaded public key from " + filename);
    }
    
    /**
     * Sets DES key (8 bytes) for encryption
     * Assumes this key is shared between sender and receiver
     */
    public void setDESKey(String key) {
        if (key.length() != 8) {
            throw new IllegalArgumentException("[INFO] DES key must be 8 characters");
        }
        this.desKey = key.getBytes(StandardCharsets.UTF_8);
    }
    
    /**
     * Signs data using SHA1withRSA
     * 
     * @param data Data to sign
     * @return Digital signature
     */
    public byte[] sign(String data) throws Exception {
        if (privateKey == null) {
            throw new IllegalStateException("[ERR] No private key loaded!");
        }
        
        byte[] dataBytes = data.getBytes(StandardCharsets.UTF_8);
        
        Signature signature = Signature.getInstance("SHA1withRSA");
        signature.initSign(privateKey);
        signature.update(dataBytes);
        
        byte[] signatureBytes = signature.sign();
        
        System.out.println("Signed message (length: " + dataBytes.length + " bytes)");
        System.out.println("Signature (length: " + signatureBytes.length + " bytes)");
        
        return signatureBytes;
    }
    
    /**
     * Verifies signature using SHA1withRSA
     * 
     * @param data Original data
     * @param signatureBytes Digital signature
     * @return true if signature is valid
     */
    public boolean verifySignature(String data, byte[] signatureBytes) throws Exception {
        if (publicKey == null) {
            throw new IllegalStateException("[ERR] No public key loaded!");
        }
        
        byte[] dataBytes = data.getBytes(StandardCharsets.UTF_8);
        
        Signature signature = Signature.getInstance("SHA1withRSA");
        signature.initVerify(publicKey);
        signature.update(dataBytes);
        
        boolean isValid = signature.verify(signatureBytes);
        
        if (isValid) {
            System.out.println("[PASS] Signature is valid!");
        } else {
            System.out.println("[FAIL] Signature is invalid!");
        }
        
        return isValid;
    }
    
    /**
     * Sign-and-Encrypt: Signs data, then encrypts (data + signature) with DES
     * 
     * Format: [4 bytes: data_length][4 bytes: sig_length][data][signature]
     * All encrypted with DES/CBC/PKCS5Padding
     * Output: [8 bytes: IV][encrypted_content]
     */
    public void signAndEncrypt(String data, String outputFile) throws Exception {
        if (privateKey == null) {
            throw new IllegalStateException("[ERR] No private key loaded!");
        }
        if (desKey == null) {
            throw new IllegalStateException("[ERR] DES key not set!");
        }
        
        System.out.println("\n=== SIGN AND ENCRYPT ===");
        
        byte[] dataBytes = data.getBytes(StandardCharsets.UTF_8);
        
        // Step 1: Sign data
        System.out.println("Step 1: Signing data...");
        byte[] signatureBytes = sign(data);
        
        // Step 2: Create package (data + signature)
        System.out.println("Step 2: Creating data package...");
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        DataOutputStream dos = new DataOutputStream(baos);
        
        dos.writeInt(dataBytes.length);
        dos.writeInt(signatureBytes.length);
        dos.write(dataBytes);
        dos.write(signatureBytes);
        dos.flush();
        
        byte[] packageBytes = baos.toByteArray();
        
        // Step 3: Encrypt with DES
        System.out.println("Step 3: Encrypting with DES...");
        SecretKeySpec secretKey = new SecretKeySpec(desKey, "DES");
        Cipher cipher = Cipher.getInstance("DES/CBC/PKCS5Padding");
        cipher.init(Cipher.ENCRYPT_MODE, secretKey);
        
        byte[] iv = cipher.getIV();
        byte[] ciphertext = cipher.doFinal(packageBytes);
        
        // Save to file (IV + ciphertext)
        try (FileOutputStream fos = new FileOutputStream(outputFile)) {
            fos.write(iv);
            fos.write(ciphertext);
        }
        
        System.out.println("[SUCCESS] Sign and encrypt completed, saved to " + outputFile);
        System.out.println("  - Original data: " + dataBytes.length + " bytes");
        System.out.println("  - Signature: " + signatureBytes.length + " bytes");
        System.out.println("  - Encrypted file: " + (iv.length + ciphertext.length) + " bytes");
    }
    
    /**
     * Decrypt-and-Verify: Decrypts data, then verifies signature
     * 
     * @param inputFile Encrypted file
     * @return Decrypted data if signature is valid
     */
    public String decryptAndVerify(String inputFile) throws Exception {
        if (publicKey == null) {
            throw new IllegalStateException("[ERR] No public key loaded!");
        }
        if (desKey == null) {
            throw new IllegalStateException("[ERR] DES key not set!");
        }
        
        System.out.println("\n=== DECRYPT AND VERIFY ===");
        
        // Step 1: Decrypt with DES
        System.out.println("Step 1: Decrypting with DES...");
        byte[] fileBytes = Files.readAllBytes(Paths.get(inputFile));
        
        byte[] iv = new byte[8];
        System.arraycopy(fileBytes, 0, iv, 0, 8);
        
        byte[] ciphertext = new byte[fileBytes.length - 8];
        System.arraycopy(fileBytes, 8, ciphertext, 0, ciphertext.length);
        
        SecretKeySpec secretKey = new SecretKeySpec(desKey, "DES");
        Cipher cipher = Cipher.getInstance("DES/CBC/PKCS5Padding");
        IvParameterSpec ivSpec = new IvParameterSpec(iv);
        cipher.init(Cipher.DECRYPT_MODE, secretKey, ivSpec);
        
        byte[] packageBytes = cipher.doFinal(ciphertext);
        
        // Step 2: Extract data and signature
        System.out.println("Step 2: Extracting data and signature...");
        ByteArrayInputStream bais = new ByteArrayInputStream(packageBytes);
        DataInputStream dis = new DataInputStream(bais);
        
        int dataLength = dis.readInt();
        int sigLength = dis.readInt();
        
        byte[] dataBytes = new byte[dataLength];
        byte[] signatureBytes = new byte[sigLength];
        
        dis.readFully(dataBytes);
        dis.readFully(signatureBytes);
        
        String data = new String(dataBytes, StandardCharsets.UTF_8);
        
        System.out.println("  - Data: " + dataLength + " bytes");
        System.out.println("  - Signature: " + sigLength + " bytes");
        
        // Step 3: Verify signature
        System.out.println("Step 3: Verifying signature...");
        boolean isValid = verifySignature(data, signatureBytes);
        
        if (!isValid) {
            throw new SecurityException("[FAIL] Signature verification failed!");
        }
        
        return data;
    }
    
    /**
     * Interactive main menu
     */
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        Lab05_3 ds = new Lab05_3();
        
        System.out.println("============================================================");
        System.out.println("DIGITAL SIGNATURE PROGRAM (SHA1withRSA)");
        System.out.println("============================================================");
        System.out.println("WARNING: SHA-1 is deprecated. For educational use only!");
        
        try {
            System.out.println("\nChoose function:");
            System.out.println("1. Generate RSA key pair");
            System.out.println("2. Sign message");
            System.out.println("3. Verify signature");
            System.out.println("4. Sign and encrypt (signAndEncrypt)");
            System.out.println("5. Decrypt and verify (decryptAndVerify)");
            System.out.println("6. Demo basic sign/verify");
            System.out.println("7. Demo sign-and-encrypt");
            
            System.out.print("\nChoice (1-7): ");
            String choice = scanner.nextLine().trim();
            
            switch (choice) {
                case "1":
                    // Generate keys
                    System.out.print("Enter key size (default 2048): ");
                    String sizeStr = scanner.nextLine().trim();
                    int keySize = sizeStr.isEmpty() ? 2048 : Integer.parseInt(sizeStr);
                    
                    ds.generateRSAKeys(keySize);
                    
                    System.out.print("Enter base name for key files: ");
                    String baseName = scanner.nextLine().trim();
                    if (baseName.isEmpty()) {
                        baseName = "signature_key";
                    }
                    
                    ds.saveRSAKeys(baseName);
                    break;
                    
                case "2":
                    // Sign
                    System.out.print("Enter private key file: ");
                    String privKey = scanner.nextLine().trim();
                    
                    if (!new File(privKey).exists()) {
                        System.out.println("File " + privKey + " does not exist!");
                        return;
                    }
                    
                    ds.loadPrivateKey(privKey);
                    
                    System.out.print("Enter message to sign: ");
                    String message = scanner.nextLine().trim();
                    
                    byte[] signature = ds.sign(message);
                    
                    System.out.print("Enter file to save signature (default: signature.bin): ");
                    String sigFile = scanner.nextLine().trim();
                    if (sigFile.isEmpty()) {
                        sigFile = "signature.bin";
                    }
                    
                    try (FileOutputStream fos = new FileOutputStream(sigFile)) {
                        fos.write(signature);
                    }
                    System.out.println("Signature saved to " + sigFile);
                    break;
                    
                case "3":
                    // Verify
                    System.out.print("Enter public key file: ");
                    String pubKey = scanner.nextLine().trim();
                    
                    if (!new File(pubKey).exists()) {
                        System.out.println("File " + pubKey + " does not exist!");
                        return;
                    }
                    
                    ds.loadPublicKey(pubKey);
                    
                    System.out.print("Enter original message: ");
                    String origMessage = scanner.nextLine().trim();
                    
                    System.out.print("Enter signature file: ");
                    String sigFileVerify = scanner.nextLine().trim();
                    
                    if (!new File(sigFileVerify).exists()) {
                        System.out.println("File " + sigFileVerify + " does not exist!");
                        return;
                    }
                    
                    byte[] signatureBytes = Files.readAllBytes(Paths.get(sigFileVerify));
                    ds.verifySignature(origMessage, signatureBytes);
                    break;
                    
                case "4":
                    // Sign and encrypt
                    System.out.print("Enter private key file: ");
                    String privKey2 = scanner.nextLine().trim();
                    
                    if (!new File(privKey2).exists()) {
                        System.out.println("File " + privKey2 + " does not exist!");
                        return;
                    }
                    
                    ds.loadPrivateKey(privKey2);
                    
                    System.out.print("Enter DES key (8 characters): ");
                    String desKey = scanner.nextLine().trim();
                    ds.setDESKey(desKey);
                    
                    System.out.print("Enter message: ");
                    String msg = scanner.nextLine().trim();
                    
                    System.out.print("Enter output file (default: encrypted.bin): ");
                    String outFile = scanner.nextLine().trim();
                    if (outFile.isEmpty()) {
                        outFile = "encrypted.bin";
                    }
                    
                    ds.signAndEncrypt(msg, outFile);
                    break;
                    
                case "5":
                    // Decrypt and verify
                    System.out.print("Enter public key file: ");
                    String pubKey2 = scanner.nextLine().trim();
                    
                    if (!new File(pubKey2).exists()) {
                        System.out.println("File " + pubKey2 + " does not exist!");
                        return;
                    }
                    
                    ds.loadPublicKey(pubKey2);
                    
                    System.out.print("Enter DES key (8 characters): ");
                    String desKey2 = scanner.nextLine().trim();
                    ds.setDESKey(desKey2);
                    
                    System.out.print("Enter encrypted file: ");
                    String inFile = scanner.nextLine().trim();
                    
                    if (!new File(inFile).exists()) {
                        System.out.println("File " + inFile + " does not exist!");
                        return;
                    }
                    
                    String decryptedData = ds.decryptAndVerify(inFile);
                    
                    System.out.println("\n--- RESULT ---");
                    System.out.println("Decrypted data: " + decryptedData);
                    break;
                    
                case "6":
                    demoBasicSignVerify();
                    break;
                    
                case "7":
                    demoSignAndEncrypt();
                    break;
                    
                default:
                    System.out.println("[ERR] Invalid choice!");
            }
            
        } catch (Exception e) {
            System.out.println("\nError: " + e.getMessage());
            e.printStackTrace();
        } finally {
            scanner.close();
        }
    }
    
    /**
     * Demo: Basic sign and verify
     */
    private static void demoBasicSignVerify() throws Exception {
        System.out.println("\n============================================================");
        System.out.println("DEMO: BASIC SIGN AND VERIFY");
        System.out.println("============================================================");
        
        Lab05_3 ds = new Lab05_3();
        ds.generateRSAKeys(2048);
        
        String message = "This is a test message for digital signature!";
        System.out.println("\nMessage: " + message);
        
        // Sign
        System.out.println("\n--- Signing message ---");
        byte[] signature = ds.sign(message);
        
        // Verify (correct)
        System.out.println("\n--- Verifying signature (original message) ---");
        ds.verifySignature(message, signature);
        
        // Verify (tampered)
        System.out.println("\n--- Verifying signature (tampered message) ---");
        String tamperedMessage = "This message has been TAMPERED!";
        ds.verifySignature(tamperedMessage, signature);
    }
    
    /**
     * Demo: Sign and encrypt complete workflow
     */
    private static void demoSignAndEncrypt() throws Exception {
        System.out.println("\n============================================================");
        System.out.println("DEMO: SIGN AND ENCRYPT WORKFLOW");
        System.out.println("============================================================");
        
        // Sender
        System.out.println("\n--- SENDER ---");
        Lab05_3 sender = new Lab05_3();
        sender.generateRSAKeys(2048);
        sender.setDESKey("sharedKY");  // Shared DES key
        
        String message = "Secret message that needs confidentiality and authenticity!";
        System.out.println("Message: " + message);
        
        sender.signAndEncrypt(message, "encrypted_message.bin");
        
        // Receiver
        System.out.println("\n--- RECEIVER ---");
        Lab05_3 receiver = new Lab05_3();
        receiver.publicKey = sender.publicKey;  // Received public key from sender
        receiver.setDESKey("sharedKY");  // Shared DES key
        
        String decryptedData = receiver.decryptAndVerify("encrypted_message.bin");
        
        System.out.println("\n--- RESULT ---");
        System.out.println("Received message: " + decryptedData);
        System.out.println("Message authenticity: VERIFIED");
        
        // Clean up
        new File("encrypted_message.bin").delete();
    }
}
